File: GeneralBrokenLines/fortran/gbltst.f90
!

! Code converted using TO_F90 by Alan Miller
! Date: 2012-03-03  Time: 17:00:22

!> \file
!! Test program.

!> Simple tests with GBL.
!!
!! Runs with gcc4 (for RANDOM_NUMBER) or gcc3 (simple uniform distribution).
!!
!! Simple track model:
!!      - initial direction (COSL,0.,SINL)
!!      - solenoidal magnectic field B=(0,0,Bz)
!!      - simplified jacobian (parabolic in arclength)
!!      - distortions in curvilinear system
!!      - multiple scattering

PROGRAM test
    use gbltraj, only: gblini, gbladp, gbladm, gblads, gbladl, gbladg, gbladx, &
                       gblfit, gblres, gblmp2

    PARAMETER (nloc=2) ! number of 'local' parameters
    PARAMETER (mp=5+nloc)
    PARAMETER (mp2=(mp+1)*mp/2)
    DOUBLE PRECISION :: dpar(mp), dcov(mp2), dpseed(mp2), djac(5,5),daux(15),  &
    ajacp,ajacn,pm2l,pl2m,det

    DIMENSION clpar(5),dirt(3),dirm(3,2),tmp(5),sarc(100),  &
    dif(2),sig(2),prec(2),ajacp(5,5),ajacn(5,5),  &
    dirz(3),diru(3),dirv(3),pm2l(2,2),pl2m(2,2),beta0(2),  &
    clerr(5),lder(nloc),derlc(2,nloc),thsig(2),thprec(2)

    DATA dirz / 0.0, 0.0, 1.0 / ! Z direction


    ! (type of) local system

    !     ITYPE=0 ! curvilinear track parameter (q/p,lambda,phi,x_t,y_t)
    itype=1 ! curvilinear local system    (q/p,v',w',v,w), (v,w)=(x_t,y_t)
    !     ITYPE=2 ! 'magnetic'  local system    (q/p,v',w',v,w), (v,w)=(x_t,z)
    !               (perpendicular, parallel to B)
    PRINT *, ' GBLTST ITYPE: ', itype

    bz=1.0 ! magnetic field
    bfac=bz*0.2998
    !     initial track direction (from prefit)
    sinl=0.3               ! sin(lambda)
    cosl=SQRT(1.0-sinl**2) ! cos(lambla)
    phi=0.                 ! move in X direction
    !     preset to unit matrix
    DO i=1,5
        DO j=1,5
            ajacp(i,j)=0.0D0
            ajacn(i,j)=0.0D0
        END DO
        ajacp(i,i)=1.0D0
        ajacn(i,i)=1.0D0
    END DO
    beta0(1)=0.0
    beta0(2)=0.0
    !     distortions to track parameters
    clerr(1)= 0.001
    clerr(2)=-0.1
    clerr(3)= 0.2
    clerr(4)=-0.15
    clerr(5)= 0.25

    ip=0
    DO i=1,mp
        DO j=1,i
            dpseed(ip+j)=0.0D0
        END DO
        ip=ip+i
        IF (i <= 5) dpseed(ip)=1.0D0/DBLE(clerr(i)**2)
    END DO
    !      print *, ' DPSEED '
    !      CALL DBPRV(6,DPSEED,MP)

    ntr=10000 ! number of tracks
    schi2=0.0
    DO itr=1,ntr
        n=10
        s=0.0            ! at vertex
        s0=10.           ! first measurement
        step1=1.0/cosl   ! constant steps
        step2=2.0/cosl   !    in RPhi
        dert0=1.0
        !       curvilinear distortions d(q/p,theta,phi,xt,yt)
        DO k=1,5
            !          CLPAR(K)=0.0
            !          CLPAR(K)=CLERR(K)
            clpar(k)=clerr(k)*unrm()
        END DO
  
        ipnt1=0        ! first measurement, distortions defined here
  
        theta0=1.0E-3  ! 1 mrad multiple scatt.
        IF (itype == 1) THEN
            thsig(1)=theta0         ! MS error in v' (dx_t/dz_t)
            thsig(2)=theta0         ! MS error in w' (dy_t/dz_t)
        ELSE IF (itype == 2) THEN
            thsig(1)=theta0/cosl    ! MS error in v' (dx_t/ds_2D)
            thsig(2)=theta0/cosl**2 ! MS error in w' (dz/ds_2D)
        ELSE
            thsig(1)=theta0         ! MS error in lambda
            thsig(2)=theta0/cosl    ! MS error in phi
        END IF
        thprec(1)=1.0/thsig(1)**2 ! diagonal of inverse
        thprec(2)=1.0/thsig(2)**2 ! covariance matrix
  
        sig(1)=1.0E-3         ! 10 mu resolution
        sig(2)=1.0E-3         ! 10 mu resolution
        prec(1)=1.0/sig(1)**2 ! diagonal of inverse
        prec(2)=1.0/sig(2)**2 ! covariance matrix
        CALL gblini(1)
        !        CALL GBLINP(1,1)
        !        point at vertex
        !        CALL gbladp(ajacn,ipnt)
        !        sarc(ipnt)=s
        !        CALL gbltjc(itype,s0,bfac,cosl,ajacn)

        ! create track
        s=s0
  
        DO i=1,n
            !         print *, ' measurement ', I, S, OFF, SLP
            ! track direction
            cphi=COS(phi)
            sphi=SIN(phi)
            dirt(1)= cosl*cphi
            dirt(2)= cosl*sphi
            dirt(3)= sinl
            ! U = Z x T / |Z x T|
            diru(1)=-sphi
            diru(2)= cphi
            diru(3)= 0.
            ! V = T x U
            dirv(1)=-sinl*cphi
            dirv(2)=-sinl*sphi
            dirv(3)= cosl
            ! V = Z
            IF (itype == 2) THEN
                dirv(1)=0.
                dirv(2)=0.
                dirv(3)=1.
            END IF
            ! RPhi meas. ( = Y as track moves in X direction)
            eps=0.
            IF (MOD(i,2) == 0) eps=0.5
            dirm(1,1)=eps ! direction of measurement
            dirm(2,1)=SQRT(1.0-eps**2)
            dirm(3,1)=0.0
            ! Z meas.
            eps=0.
            !          IF (MOD(I,2).EQ.0) EPS=0.3
            dirm(1,2)=0.0 ! direction of measurement
            dirm(2,2)=eps
            dirm(3,2)=SQRT(1.0-eps**2)
            ! projection (du/dm: measurement to local (curvilinear))
            DO l=1,2
                pm2l(l,1)= DBLE(dirm(1,l)*diru(1)+dirm(2,l)*diru(2)+dirm(3,l)*diru(3))
                pm2l(l,2)= DBLE(dirm(1,l)*dirv(1)+dirm(2,l)*dirv(2)+dirm(3,l)*dirv(3))
            END DO
            ! projection (dm/du: local to measurement)
            det=pm2l(1,1)*pm2l(2,2)-pm2l(1,2)*pm2l(2,1)
            pl2m(1,1)= pm2l(2,2)/det
            pl2m(1,2)=-pm2l(1,2)/det
            pl2m(2,1)=-pm2l(2,1)/det
            pl2m(2,2)= pm2l(1,1)/det
            ! measurement - prediction in measurement system with error
            DO m=1,2
                dif(m)=clpar(4)*SNGL(pl2m(m,1)) +clpar(5)*SNGL(pl2m(m,2))+sig(m)*unrm()
            END DO
            ! distort with local parameters
            !          DIF(1)=DIF(1)+DERT0*0.0075
            !          DIF(2)=DIF(2)-DERT0*0.0025
    
            CALL gbladp(ajacn,ipnt)
            sarc(ipnt)=s
            CALL gbladm(pl2m,dif,prec)
            IF (ipnt1 == 0) ipnt1=ipnt
            ! local or global parameters
            lder(1)=4711
            lder(2)=4712
            derlc(1,1)= dert0
            derlc(2,1)=0.0
            derlc(1,2)=0.0
            derlc(2,2)=-dert0
            !          CALL GBLADL(2,DERLC,IRET)
            !          CALL GBLADG(2,LDER,DERLC,IRET)
            dert0=-dert0
            ! propagate by STEP1
            ds=step1
            CALL gbltjc(itype,step1,bfac,cosl,ajacn)
    
            DO k=1,5
                tmp(k)=clpar(k)
            END DO
            DO k=1,5
                clpar(k)=0.0
                DO l=1,5
                    clpar(k)=clpar(k)+SNGL(ajacn(k,l))*tmp(l)
                END DO
            END DO
            s=s+step1
            IF (i < n) THEN
                CALL gbladp(ajacn,ipnt)
                sarc(ipnt)=s
                CALL gblads(beta0,thprec)
            END IF
            ! scatter a little
            clpar(2)=clpar(2)+thsig(1)*unrm()
            clpar(3)=clpar(3)+thsig(2)*unrm()
            ! propagate by STEP1
            CALL gbltjc(itype,step2,bfac,cosl,ajacn)

            DO k=1,5
                tmp(k)=clpar(k)
            END DO
            DO k=1,5
                clpar(k)=0.0
                DO l=1,5
                    clpar(k)=clpar(k)+SNGL(ajacn(k,l))*tmp(l)
                END DO
            END DO
            s=s+step2
        END DO
  
        npnt=ipnt
               CALL GBLADX(IPNT1,DPSEED)
  
        !        CALL GBLDMP
        CALL gblfit('',mrank,np,ndf,chi2,wls)
        !        print *, ' GBLFIT ', MRANK, NP, NDF, CHI2, WLS
        schi2=schi2+chi2/FLOAT(ndf)
        !        CALL GBLFIT('HT',MRANK,NDF,CHI2,WLS)
        IF (itr <= 1) THEN
            PRINT *, ' GBLFIT ', mrank, np, ndf, chi2, wls
            CALL gblres(ipnt1,dpar,dcov)
            PRINT *, ' DPAR ', (dpar(k),k=1,np)
            PRINT *, ' DCOV(loc) '
            CALL dbprv(6,dcov,np)
            CALL gbll2c(itype,cosl,djac)
            CALL dbavat(dcov,djac,daux,5,5)
            PRINT *, ' DCOV(cvl) '
            CALL dbprv(6,daux,5)
        END IF
  
    !        CALL GBLMP2(IRET) ! prepare input for Millepede-II
  
    END DO

    schi2=schi2/FLOAT(ntr)
    PRINT *, ' === <Chi2/ndf> === ', ntr, schi2

END PROGRAM test

!> Simplified jacobian (quadratic in DS)

SUBROUTINE gbltjc(itype,ds,bfac,cosl,ajac)


    INTEGER, INTENT(IN)                      :: itype
    REAL, INTENT(IN)                         :: ds
    REAL, INTENT(IN)                         :: bfac
    REAL, INTENT(IN OUT)                     :: cosl
    DOUBLE PRECISION, INTENT(OUT)            :: ajac(5,5)


    IF (itype == 1) THEN
        !       curvilinear local system (q/p,v',w',v,w), (v,w)=(x_t,y_t)
        ajac(1,1)= 1.0D0
        ajac(2,1)=-DBLE(bfac*ds*cosl)
        ajac(2,2)= 1.0D0
        ajac(3,3)= 1.0D0
        ajac(4,1)=-DBLE(0.5*bfac*ds*ds*cosl)
        ajac(4,2)= DBLE(ds)
        ajac(4,4)= 1.0D0
        ajac(5,3)= DBLE(ds)
        ajac(5,5)= 1.0D0
    ELSE IF (itype == 2) THEN
        !       local system (q/p,v',w',v,w), (v,w)=(x_t,z)
        ajac(1,1)= 1.0D0
        ajac(2,1)=-DBLE(bfac*ds)
        ajac(2,2)= 1.0D0
        ajac(3,3)= 1.0D0
        ajac(4,1)=-DBLE(0.5*bfac*ds*ds*cosl)
        ajac(4,2)= DBLE(ds*cosl)
        ajac(4,4)= 1.0D0
        ajac(5,3)= DBLE(ds*cosl)
        ajac(5,5)= 1.0D0
    ELSE
        !       curvilinear track parameter (q/p,lambda,phi,x_t,y_t)
        ajac(1,1)= 1.0D0
        ajac(2,2)= 1.0D0
        ajac(3,1)=-DBLE(bfac*ds)
        ajac(3,3)= 1.0D0
        ajac(4,1)=-DBLE(0.5*bfac*ds*ds*cosl)
        ajac(4,3)= DBLE(ds*cosl)
        ajac(4,4)= 1.0D0
        ajac(5,2)= DBLE(ds)
        ajac(5,5)= 1.0D0
    END IF

    RETURN
END SUBROUTINE gbltjc

!> Simplified transformation local to curvilinear

SUBROUTINE gbll2c(itype,cosl,djac)


    INTEGER, INTENT(IN)                      :: itype
    REAL, INTENT(IN OUT)                     :: cosl
    DOUBLE PRECISION, INTENT(OUT)            :: djac(5,5)

    !     preset with unit matrix
    DO k=1,5
        DO l=1,5
            djac(k,l)=0.0D0
        END DO
        djac(k,k)=1.0D0
    END DO

    IF (itype == 1) THEN
        !       curvilinear local system (q/p,v',w',v,w), (v,w)=(x_t,y_t)
        djac(2,2)=0.0D0
        djac(2,3)=1.0D0/DBLE(cosl)
        djac(3,2)=1.0D0
        djac(3,3)=0.0D0
    ELSE IF (itype == 2) THEN
        !       local system (q/p,v',w',v,w), (v,w)=(x_t,z)
        djac(2,2)=0.0D0
        djac(2,3)=1.0D0
        djac(3,2)=DBLE(cosl)**2
        djac(3,3)=0.0D0
        djac(5,5)=DBLE(cosl)
    END IF

    RETURN
END SUBROUTINE gbll2c

!> Simple unit normal
FUNCTION unrm()
    DIMENSION r(12)
    !      DATA IX / 4711 /      ! use for gcc3
    CALL random_number(r) ! comment out for gcc3
    unrm=-6.0
    DO k=1,12
        !      R(K)=UNIF(IX)         ! use for gcc3
        unrm=unrm+r(k)
    END DO
END

!> Simple uniform random numbers
FUNCTION unif(ix)
    ! portable random number generator using the recursion:
    ! IX = 16807 * IX MOD (2**31-1)
    k1 = ix/127773
    ix = 16807* (ix - k1*127773) - k1 * 2836
    IF (ix < 0) ix = ix + 2147483647
    unif = ix*4.656612875E-10
    RETURN
END FUNCTION unif

File: GeneralBrokenLines/python/gbltst.py
'''
Simple Test Program for General Broken Lines.

Created on Jul 27, 2011

@author: kleinwrt
'''

## \file
# examples
#
# \author Claus Kleinwort, DESY, 2011 (Claus.Kleinwort@desy.de)
#
#  \copyright
#  Copyright (c) 2011 - 2016 Deutsches Elektronen-Synchroton,
#  Member of the Helmholtz Association, (DESY), HAMBURG, GERMANY \n\n
#  This library is free software; you can redistribute it and/or modify
#  it under the terms of the GNU Library General Public License as
#  published by the Free Software Foundation; either version 2 of the
#  License, or (at your option) any later version. \n\n
#  This library is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU Library General Public License for more details. \n\n
#  You should have received a copy of the GNU Library General Public
#  License along with this program (see the file COPYING.LIB for more
#  details); if not, write to the Free Software Foundation, Inc.,
#  675 Mass Ave, Cambridge, MA 02139, USA.

import numpy as np
import math
import time
from gblfit import GblPoint, GblTrajectory
#

## Create points on initial trajectory, create trajectory from points,
#  fit and write trajectory to MP-II binary file,
#  get track parameter corrections and covariance matrix at points.
#  
#  Equidistant measurement layers and thin scatterers, propagation 
#  with simple jacobian (quadratic in arc length differences).
#  Curvilinear system (T,U,V) or measurement system (I,J,K) as local
#  coordinate system and (q/p, slopes, offsets) as local track parameters.
#  
#  This example simulates and refits tracks in a system of planar detectors
#  with 2D measurements in a constant magnet field in Z direction using
#  the curvilinear system as local system. The true track parameters are
#  randomly smeared with respect to a (constant and straight) reference
#  trajectory with direction (lambda, phi) and are used (only) for the
#  on-the-fly simulation of the measurements and scatterers. The predictions
#  from the reference trajectory are therefore always zero and the residuals
#  needed (by addMeasurement) are equal to the measurements.
#
def example1():

#
  np.random.seed(47117)

  nTry = 1000  #: number of tries
  nLayer = 5  #: number of detector layers
  print " Gbltst $Rev: 116 $ ", nTry, nLayer
  start = time.clock()
# track direction
  sinLambda = 0.3
  cosLambda = math.sqrt(1.0 - sinLambda ** 2)
  sinPhi = 0.
  cosPhi = math.sqrt(1.0 - sinPhi ** 2)
# Curvilinear system: track direction T, U = Z x T / |Z x T|, V = T x U
  tuvDir = np.array([[cosLambda * cosPhi, cosLambda * sinPhi, sinLambda], \
                     [-sinPhi, cosPhi, 0.], \
                     [-sinLambda * cosPhi, -sinLambda * sinPhi, cosLambda]])
# measurement resolution
  measErr = np.array([ 0.001, 0.001])  # 10 mu
  measPrec = 1.0 / measErr ** 2
# scattering error
  scatErr = 0.001  # 1 mread
# RMS of CurviLinear track parameters (Q/P, slopes, offsets)
  clErr = np.array([0.001, -0.1, 0.2, -0.15, 0.25])
  # precision matrix for external seed (in local system)
  locSeed = None
  seedLabel = 0  # label of point with seed
  if seedLabel != 0:
    print " external seed at label ", seedLabel
#
  bfac = 0.2998  # Bz*c for Bz=1
  step = 1.5 / cosLambda  # constant steps in RPhi
#
  Chi2Sum = 0.
  NdfSum = 0
  LostSum = 0.
#
  binaryFile = open("milleBinaryISN.dat", "wb")
#
  for iTry in range(nTry):
# generate (CurviLinear) track parameters 
    clNorm = np.random.normal(0., 1., 5)  
    clPar = clErr * clNorm
    # covariance matrix
    clCov = np.eye(5)
    for i in range(5):
      clCov[i, i] = clErr[i] ** 2
# arclength
    s = 0.
# point-to-point jacobian (from previous point)    
    jacPointToPoint = np.eye(5)
# additional (local or global) derivatives    
    addDer = np.array([[1.0], [0.0]])
    labGlobal = np.array([[4711], [4711]])
# create trajectory
    traj = GblTrajectory(bfac != 0.)

# at previous point: transformation from local to curvilinear system 
    oldL2c = np.eye(5)
   
    for iLayer in range(nLayer):
      #print " layer ", iLayer
# measurement directions (J,K) from stereo angle
      sinStereo = (0. if iLayer % 2 == 0 else 0.1) 
      cosStereo = math.sqrt(1.0 - sinStereo ** 2)    
# measurement system: I, J ,K    
      ijkDir = np.array([[1., 0., 0.], \
                         [0., cosStereo, sinStereo], \
                         [0., -sinStereo, cosStereo]])
# local system: measurement or curvilinear 
      #local = gblMeasSystem(ijkDir, tuvDir)
      local = gblCurviSystem(ijkDir, tuvDir)
# projections
      proL2m = local.getTransLocalToMeas()
      proL2c = local.getTransLocalToCurvi()
      proC2l = np.linalg.inv(proL2c)
# projection curvilinear to measurement directions
      proC2m = local.getTransCurviToMeas()
# measurement - prediction in measurement system with error
      measNorm = np.random.normal(0., 1., 2)  
      meas = np.dot(proC2m, clPar[3:5]) + measErr * measNorm
# jacobian is calculated in curvilinear system and transformed
      jac = np.dot(proC2l, np.dot(jacPointToPoint, oldL2c))
# point with (independent) measurements (in measurement system)
      point = GblPoint(jac)
      point.addMeasurement([proL2m, meas, measPrec])
# additional local parameters?
#      point.addLocals(addDer)
# additional global parameters?
      point.addGlobals(labGlobal, addDer)
      addDer = -addDer  # locDer flips sign every measurement      
# add point to trajectory      
      iLabel = traj.addPoint(point)
      if iLabel == abs(seedLabel):
        clSeed = np.linalg.inv(clCov)
        locSeed = np.dot(proL2c, np.dot(clSeed, proL2c.T))
# propagate to scatterer
      jacPointToPoint = gblSimpleJacobian(step, cosLambda, bfac)
      clPar = np.dot(jacPointToPoint, clPar)
      clCov = np.dot(jacPointToPoint, np.dot(clCov, jacPointToPoint.T))
      s += step
      if (iLayer < nLayer - 1):
        scat = np.array([0., 0.])
# point with scatterer
        jac = np.dot(proC2l, np.dot(jacPointToPoint, proL2c))
        point = GblPoint(jac)
        scatP = local.getScatPrecision(scatErr)
        point.addScatterer([scat, scatP])
        iLabel = traj.addPoint(point)
        if iLabel == abs(seedLabel):
          clSeed = np.linalg.inv(clCov)
          locSeed = np.dot(proL2c, np.dot(clSeed, proL2c.T))

# scatter a little    
        scatNorm = np.random.normal(0., 1., 2)  
        clPar[1:3] = clPar[1:3] + scatErr * scatNorm
# propagate to next measurement layer    
        clPar = np.dot(jacPointToPoint, clPar)
        clCov = np.dot(jacPointToPoint, np.dot(clCov, jacPointToPoint.T))
        s += step
      oldL2c = proL2c
 
# add external seed
    if locSeed is not None:    
      traj.addExternalSeed(seedLabel, locSeed)
# dump trajectory
#    traj.dump()
  
# fit trajectory
    Chi2, Ndf, Lost = traj.fit()
    print " Record, Chi2, Ndf, Lost", iTry, Chi2, Ndf, Lost
# write to MP binary file    
#    traj.milleOut(binaryFile)
# sum up    
    Chi2Sum += Chi2
    NdfSum += Ndf
    LostSum += Lost
# get corrections and covariance matrix at points 
    if (iTry == 0):
      for i in range(1, nLayer + 1):      
        locPar, locCov = traj.getResults(-i)
        print " >Point ", i
        print " locPar ", locPar
        #print " locCov ", locCov      
        locPar, locCov = traj.getResults(i)
        print " Point> ", i
        print " locPar ", locPar
        #print " locCov ", locCov
# check residuals        
      for i in range(traj.getNumPoints()):
        numData, aResiduals, aMeasErr, aResErr, aDownWeight = traj.getMeasResults(i + 1)
        for j in range(numData):
          print " measRes " , i, j, aResiduals[j], aMeasErr[j], aResErr[j], aDownWeight[j]   
        numData, aResiduals, aMeasErr, aResErr, aDownWeight = traj.getScatResults(i + 1)
        for j in range(numData):
          print " scatRes " , i, j, aResiduals[j], aMeasErr[j], aResErr[j], aDownWeight[j]   
#
  end = time.clock()
  print " Time [s] ", end - start
  print " Chi2Sum/NdfSum ", Chi2Sum / NdfSum
  print " LostSum/nTry ", LostSum / nTry

## Read trajectory from MP-II binary file and refit.
#
def example2():
#  
  binaryFile = open("milleBinaryISN.dat", "rb")
  nRec = 0
  maxRec = 10  #: maximum number of records to read
  Chi2Sum = 0.
  NdfSum = 0
  LostSum = 0.
  start = time.clock()
  
  try:
    while(nRec < maxRec):
# create trajectory
      traj = GblTrajectory(0)
# read from file      
      traj.milleIn(binaryFile)  # get data blocks from file
      nRec += 1
# fit trajectory      
      Chi2, Ndf, Lost = traj.fit()
      print " Record, Chi2, Ndf, Lost", nRec, Chi2, Ndf, Lost
# sum up      
      Chi2Sum += Chi2
      NdfSum += Ndf
      LostSum += Lost 
         
  except EOFError:
    pass    
  
  print " records read ", nRec
  end = time.clock()
  print " Time [s] ", end - start
  print " Chi2Sum/NdfSum ", Chi2Sum / NdfSum
  print " LostSum/nTry ", LostSum / nRec
 
## Simple jacobian.
#  
#  Simple jacobian for (q/p, slopes, offsets) in curvilinear system
#  quadratic in arc length difference.
#
#  @param ds arc length difference; float
#  @param cosl cos(lambda); float
#  @param bfac Bz*c; float
#  @return jacobian to move by 'ds' on trajectory matrix(float)
#
def gblSimpleJacobian(ds, cosl, bfac):
  jac = np.eye(5)
  jac[1, 0] = -bfac * ds * cosl
  jac[3, 0] = -0.5 * bfac * ds * ds * cosl
  jac[3, 1] = ds
  jac[4, 2] = ds  
  return jac     

## Measurement system as local system
#
# In general the precision matrix for multiple scattering is not diagonal
#
class gblMeasSystem(object):
 
  ## Construct local system
  #
  #  @param  measDir (directions of) measurement system
  #  @param  curviDir (directions of) curvilinear system
  #  
  def __init__(self, measDir, curviDir):
    ## products (T,U,V) * (I,J,K)
    self.__prod = np.dot(curviDir, measDir.T)

  ## Transformation from local to measurement system 
  #
  #  @return None (systems are identical)
  #   
  def getTransLocalToMeas(self):
    return None

  ## Transformation of (q/p, slopes, offsets) from local to curvilinear system
  #
  #  @return Transformation for track parameters; 5*5 matrix
  #   
  def getTransLocalToCurvi(self):
    meas2crv = np.zeros((5, 5))
    meas2crv[0, 0] = 1.
    meas2crv[1:3, 1:3] = self.__prod[1:3, 1:3] * self.__prod[0, 0]  # (U,V)*(J,K) * T*I
    meas2crv[3:5, 3:5] = self.__prod[1:3, 1:3]  # (U,V)*(J,K)
    return meas2crv
 
  ## Transformation from curvilinear to measurement system 
  #
  #  @return Transformation for offsets; 2*2 matrix
  #   
  def getTransCurviToMeas(self):
    return np.linalg.inv(self.__prod[1:3, 1:3])  #((U,V)*(J,K))^-1
 
  ## Scattering precision matrix in local system  
  #
  #  @return Precision; 2*2 matrix
  #   
  def getScatPrecision(self, scatErr):
    c1 = self.__prod[0, 1]  # T*J
    c2 = self.__prod[0, 2]  # T*K
    fac = (1 - c1 * c1 - c2 * c2) / (scatErr * scatErr)
    scatP = np.empty((2, 2))
    scatP[0, 0] = fac * (1 - c1 * c1)
    scatP[0, 1] = fac * (-c1 * c2)
    scatP[1, 0] = fac * (-c1 * c2)
    scatP[1, 1] = fac * (1 - c2 * c2)
    return scatP       
 
## Curvilinear system as local system
#
# The precision matrix for multiple scattering (for slopes)
# is diagonal (multiple of unit matrix).
#
class gblCurviSystem(object):

  ## Construct local system
  #
  #  @param  measDir (directions of) measurement system
  #  @param  curviDir (directions of) curvilinear system
  #    
  def __init__(self, measDir, curviDir):
    ## projection curvilinear to measurement system: ((U,V)*(J,K))^-1
    self.__c2m = np.linalg.inv(np.dot(curviDir[1:3, 1:3], measDir[1:3, 1:3].T))

  ## Transformation from local to measurement system 
  #   
  #  @return Transformation for offsets; 2*2 matrix
  #   
  def getTransLocalToMeas(self):
    return self.__c2m  #((U,V)*(J,K))^-1

  ## Transformation of (q/p, slopes, offsets) from local to curvilinear system
  #
  #  @return Transformation for track parameters; 5*5 matrix
  # 
  def getTransLocalToCurvi(self):
    return np.eye(5)  # unit matrix 
 
  ## Transformation from curvilinear to measurement system 
  #   
  #  @return Transformation for offsets; 2*2 matrix
  #   
  def getTransCurviToMeas(self):
    return self.__c2m  #((U,V)*(J,K))^-1
 
  ## Scattering precision matrix in local system  
  #
  #  @return Precision diagonal; vector(2)
  #
  def getScatPrecision(self, scatErr):
    return np.array([1., 1.]) / (scatErr * scatErr)  # diagonal only     
        
# create points on initial trajectory, create trajectory from points,
# fit and write trajectory to MP-II binary file
# get track parameter corrections and covariance matrix at points
example1()
# read trajectory from MP-II binary file and refit
#example2()
